assembl/tasks/common.py

```diff
@@ -0,0 +1,321 @@
+import sys
+import os
+import os.path
+from contextlib import nested
+import base64
+import json
+
+from invoke import Task, task as base_task
+from os.path import dirname, realpath
+DEFAULT_SECTION = "DEFAULT"
+_local_file = __file__
+if _local_file.endswith('.pyc'):
+    _local_file = _local_file[:-1]
+local_code_root = dirname(dirname(realpath(_local_file)))
+
+
+def task(*args, **kwargs):
+    pre = list(kwargs.pop('pre', args))
+    pre.append(base_task(setup_ctx))
+    return base_task(pre=pre, **kwargs)
+
+
+def exists(c, path, sudo=False):
+    """
+    Return True if given path exists on the current remote host. Taken from patchwork.
+    """
+    cmd = '{}test -e "$(echo {})"'.format('sudo ' if sudo else '', path)
+    return c.run(cmd, hide=True, warn=True).ok
+
+
+def running_locally(c):
+    return c.__class__.__module__.startswith('invoke.')
+
+
+def rec_update(d1, d2):
+    if not isinstance(d1, dict):
+        return d2
+    result = dict(d1, **d2)
+    result.update({k: rec_update(v, d2.get(k, v)) for (k, v) in d1.iteritems() if isinstance(v, dict)})
+    return result
+
+
+def get_aws_account_id(c):
+    account = os.getenv("AWS_ACCOUNT_ID")
+    if account:
+        return account
+    # Attempt to fail fast on non-EC2 machines
+    if sys.platform != 'linux2' or not os.path.exists('/sys/devices/virtual/dmi/id/bios_version'):
+        return None
+    with open('/sys/devices/virtual/dmi/id/bios_version') as f:
+        platform_info = f.read()
+        if 'amazon' not in platform_info:
+            return None
+    import requests
+    try:
+        r = requests.get('http://169.254.169.254/latest/dynamic/instance-identity/document', timeout=2)
+        if r.ok:
+            return r.json()['accountId']
+    except:
+        pass
+
+
+def get_secrets_from_manager(c, aws_secret_ids, cache=True):
+    base = {}
+    if aws_secret_ids:
+        import boto3
+        secret_clients = {}
+
+        def get_secret_client(region=None):
```

Comment:
Because it's mildly abstracting our the logic in the outer function. The intent is this function to not be used outside of the for loop of the secret ids.